include/deal.II/sundials/kinsol.h

```diff
@@ -0,0 +1,684 @@
+     * set up proper solvers and preconditioners inside this function.
+     *
+     *
+     * Arguments to the function are
+     *
+     * @param[in] t  the current time
+     * @param[in] gamma  the current factor to use in the Jacobian computation
+     * @param[in] ycur  is the current $y$ vector for the current KINSOL internal step
+     * @param[in] fcur  is the current value of the implicit right-hand side at ycur,
+     *        $f_I (t_n, ypred)$.
+     *
+     *
+     * This function should return:
+     * - 0: Success
+     * - >0: Recoverable error (KINSOL will try to change its internal parameters
+     *       and attempt a new solution step)
+     * - <0: Unrecoverable error the computation will be aborted and an assertion
+     *       will be thrown.
+     */
+    std::function<int(const VectorType &ycur,
+                      const VectorType &fcur,
+                      const VectorType &rhs,
+                      VectorType &dst)> solve_jacobian_system;
+
+    /**
+     * A function object that users may supply and that is intended to return a
+     * vector whose components are the weights used by KINSOL to compute the
+     * vector norm of the solution. The implementation of this function is
+     * optional, and it is used only if implemented.
+     */
+    std::function<VectorType&()> get_solution_scaling;
+
+    /**
+     * A function object that users may supply and that is intended to return a
+     * vector whose components are the weights used by KINSOL to compute the
+     * vector norm of the function evaluation away from the solution. The
+     * implementation of this function is optional, and it is used only if
+     * implemented.
+     */
+    std::function<VectorType&()> get_function_scaling;
+
+    /**
+     * Handle KINSOL exceptions.
+     */
+    DeclException1(ExcKINSOLError, int, << "One of the SUNDIALS KINSOL internal functions "
+                   << " returned a negative error code: "
+                   << arg1 << ". Please consult SUNDIALS manual.");
+
+
+  private:
+
+    /**
+     * Throw an exception when a function with the given name is not implemented.
+     */
+    DeclException1(ExcFunctionNotProvided, std::string,
+                   << "Please provide an implementation for the function \"" << arg1 << "\"");
+
+    /**
+     * This function is executed at construction time to set the
+     * std::function above to trigger an assert if they are not
+     * implemented.
+     */
+    void set_functions_to_trigger_an_assert();
+
+    /**
+     * KINSOL configuration data.
+     */
+    AdditionalData data;
+
+    /**
+     * KINSOL memory object.
+     */
+    void *kinsol_mem;
+
+    /**
+     * KINSOL solution vector.
+     */
+    N_Vector solution;
+
+    /**
+     * KINSOL solution scale.
+     */
+    N_Vector u_scale;
+
+    /**
+     * KINSOL f scale.
+     */
+    N_Vector f_scale;
+
+#ifdef DEAL_II_WITH_MPI
```

Comment:
you should remove this guard otherwise all the rest of the interface does not work without MPI